File: src/com/anathema_roguelike/entities/characters/perks/actions/targetingstrategies/TargetEffect.java
package com.anathema_roguelike.entities.characters.perks.actions.targetingstrategies;

import com.anathema_roguelike.stats.Stat;
import com.anathema_roguelike.stats.effects.Effect;
import com.anathema_roguelike.stats.effects.HasEffect;

public abstract class TargetEffect<T extends Targetable, S extends Stat<? extends T>> implements HasEffect<Effect<T, S>> {
	
	private Class<? extends T> targetType;
	private String name;
	
	public TargetEffect(Class<? extends T> targetType, String name) {
		this.targetType = targetType;
		this.name = name;
	}
	
	public Class<? extends T> getTargetType() {
		return targetType;
	}
	
	public void applyTo(T target) {
		getEffect().ifPresent(e -> e.applyTo(target));
	}
	
	@Override
	public String toString() {
		return name;
	}
}

File: src/com/anathema_roguelike/entities/characters/perks/actions/targetingstrategies/TargetEffect.scala
/*******************************************************************************
 * Copyright (c) 2019. Carl Minden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package com.anathema_roguelike
package entities.characters.perks.actions.targetingstrategies

import com.anathema_roguelike.stats.Stat
import com.anathema_roguelike.stats.effects.{Effect, HasEffect}

import scala.reflect.ClassTag
import scala.reflect.runtime.universe._

abstract class TargetEffect[T <: Targetable : TypeTag, +S <: Stat[T]](var name: String) extends HasEffect[Effect[T, S]] {

  def getTargetType: Class[T] = typeTagToClass

  def applyTo(target: T): Unit = {
    getEffect.foreach(e => e.applyTo(target))
  }

  override def toString: String = name
}